pkg/server/testing/fake_rootfs_client.go

```diff
@@ -0,0 +1,216 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/containerd/containerd/api/services/rootfs"
+	"github.com/containerd/containerd/api/types/descriptor"
+	"github.com/containerd/containerd/api/types/mount"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+type rootfsCalledDetail struct {
+	name     string
+	argument interface{}
+}
+
+// Kind identifies the kind of snapshot.
+type Kind int
+
+// definitions of snapshot kinds
+const (
+	KindActive Kind = iota
+	KindCommitted
+)
+
+type snapshot struct {
+	Name     string // name or key of snapshot
+	Parent   string // name of parent snapshot
+	Kind     Kind   // active or committed snapshot
+	Readonly bool   // true if readonly, only valid for active
+}
+
+// FakeRootfsClient is a simple fake rootfs client, so that cri-containerd
+// can be run for testing without requiring a real containerd setup.
+type FakeRootfsClient struct {
+	sync.Mutex
+	called       []rootfsCalledDetail
+	errors       map[string]error
+	ChainIDList  map[digest.Digest]digest.Digest
+	SnapshotList map[string]snapshot
+	MountList    map[string][]*mount.Mount
+}
+
+var _ rootfs.RootFSClient = &FakeRootfsClient{}
+
+// NewFakeRootfsClient creates a FakeRootfsClient
+func NewFakeRootfsClient() *FakeRootfsClient {
+	return &FakeRootfsClient{
+		errors:       make(map[string]error),
+		ChainIDList:  make(map[digest.Digest]digest.Digest),
+		SnapshotList: make(map[string]snapshot),
+		MountList:    make(map[string][]*mount.Mount),
+	}
+}
+
+func (f *FakeRootfsClient) popError(op string) error {
+	if f.errors == nil {
+		return nil
+	}
+	err, ok := f.errors[op]
+	if ok {
+		delete(f.errors, op)
+		return err
+	}
+	return nil
+}
+
+// InjectError inject error for call
+func (f *FakeRootfsClient) InjectError(fn string, err error) {
+	f.Lock()
+	defer f.Unlock()
+	f.errors[fn] = err
+}
+
+// InjectErrors inject errors for calls
+func (f *FakeRootfsClient) InjectErrors(errs map[string]error) {
+	f.Lock()
+	defer f.Unlock()
+	for fn, err := range errs {
+		f.errors[fn] = err
+	}
+}
+
+// ClearErrors clear errors for call
+func (f *FakeRootfsClient) ClearErrors() {
+	f.Lock()
+	defer f.Unlock()
+	f.errors = make(map[string]error)
+}
+
+func resolveDiffID(digest.Digest) digest.Digest {
+	return digest.Digest("")
+}
+
+func generateChainID(layers []*descriptor.Descriptor) digest.Digest {
```

Comment:
Let's use layer digest as diffID for now, and add some comment to explain that we should be using diffID.
https://github.com/opencontainers/image-spec/blob/eb02c88ef382324a1572fa8c66f836a00d865a3f/config.md#layer-diffid